roachpb/storage: introduce localUncertaintyLimit, teach MVCC about synthetic timestamps

nvb opened the pull request:
Informs #52745.
Informs #36431.

This PR introduces a new `localUncertaintyLimit` parameter to MVCC. The PR also renames the existing `Txn.MaxTimestamp` to `Txn.GlobalUncertaintyLimit`.

`localUncertaintyLimit` is introduced alongside `GlobalUncertaintyLimit` for two reasons:
1. it allows us to stop modifying `global_uncertainty_limit` (formerly `max_timestamp`) when applying an `observed_timestamp` to a transaction when evaluating on a replica. Instead, we can keep the `localUncertaintyLimit` directly on the stack, which better represents its lifetime.
2. it allows MVCC logic to distinguish between the pre- and post-observed timestamp uncertainty interval, depending on whether a possibly-uncertain value has a standard or synthetic timestamp.

The field is defined as follows:

> LocalUncertaintyLimit is the transaction's GlobalUncertaintyLimit, reduce
> by any observed timestamp that the transaction has acquired from the node
> that holds the lease for the range that the transaction is evaluating on.
>
> The local uncertainty limit can reduce the uncertainty interval applied
> to most values on a range. This can lead to values that would otherwise
> be considered uncertain by the original global uncertainty limit to be
> considered "certainly concurrent", and thus not causally related, with
> the transaction due to observed timestamps.
>
> However, the local uncertainty limit does not apply to all values on a
> range. Specifically, values with "synthetic timestamps" must use the
> transaction's original global uncertainty limit for the purposes of
> uncertainty, because observed timestamps do not apply to values with
> synthetic timestamps.

nvb committed:
storage: remove impossible condition in pebbleMVCCScanner.seekVersion

Once we have seeked to a version, it's not possible for the current key
to be the same and for the timestamp of the current value to be above
the seek version. So this commit removes the confusing condition.

While doing so, rename a few variables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57077)
<!-- Reviewable:end -->

nvanbenschoten commented:
The more I think about this, the more I'm open to bikeshed the `observed_max_timestamp` name. Or to keep it. But feel free to poke holes in the name if you find it ugly.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/observedts/uncertainty.go, line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOzG9i80vR728RIz-5:-MNOzG9i80vR728RIz-6:bmb6u6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b2b88d7ef9f6de83dff0a61efe62bab43814dc/pkg/kv/kvserver/observedts/uncertainty.go#L15)):*
> ```Go
> import "github.com/cockroachdb/cockroach/pkg/util/hlc"
> 
> // IsUncertain deterines whether a value with the provided timestamp is
> ```

determines
___
*[pkg/roachpb/data.proto, line 361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOy_Nl2MAUoiSeqWh4:-MNOy_Nl2MAUoiSeqWh5:bdvh78z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L361)):*
> ```protobuf
>   // holds the lease for the range that the transaction is evaluating on. As
>   // such, the field is only set on the server in the context of an evaluating
>   // request on a specific range.
> ```

Will this field ever be set in a stored `Transaction` proto, or used in an RPC?
Or is this just a convenient place to avoid plumbing through an extra parameter across function calls?
___
*[pkg/storage/pebble_mvcc_scanner.go, line 656 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOxv0e8obXAfymg7V1:-MNOxv0e8obXAfymg7V2:bncciar) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ef06227291b6e1ac80cc595db06cefff2b3c5c5/pkg/storage/pebble_mvcc_scanner.go#L656)):*
> ```Go
> 		return p.advanceKeyAtNewKey(origKey)
> 	}
> 	if p.curKey.Timestamp.LessEq(ts) {
> ```

is it worth keeping this check as an assertion, in case there is data corruption?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
I've only read parts of this PR. Flushing a couple of more comments.

> Reviewed 1 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MNPTPY927FDkwaMfwCm:bibyn1k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPQ2iW6yTT6XxrkSZw:-MNPQ2iW6yTT6XxrkSZx:b-w33y8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L397)):*
> ```Go
> 		// Note that if we own the intent (i.e. we're reading transactionally)
> 		// we want to read the intent regardless of our read timestamp and fall
> 		// into case 8 below.
> ```

Since !otherIntentVisible, we have `!p.checkUncertainty || !p.isUncertainValue(metaTS)`. I am trying to understand this if-block. It would be executed when `!p.isUncertainValue(metaTS)`. Ah, I guess there could be committed values that are `< MaxTimestamp`. Could you expand a bit on the code comment to explain this.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 656 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPSC8y606x0t_RAN7-:-MNPSC8zBahIY2dyUaSR:bturxnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L656)):*
> ```Go
> 				return p.addAndAdvance(p.curRawKey, p.curValue)
> 			}
> 			// Iterate through uncertainty interval.
> ```

IIUC, this is needed because not everything in the interval (p.ts, seekTS] is an uncertainty error, since are doing seeks based on the worst-case uncertainty, because we need to look at the synthetic timestamp bit on what we find, to determine uncertainty. Could you add a longer comment here.

The loop below could also use a comment.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
This looks good. Truth be told I didn't scrutinize the test cases, but gave a close review of the rest. Thanks for the clean commit history, that helped. The MVCC changes can definitely benefit from extensive commentary; the mvcc scanner is super dense.

> Reviewed 31 of 31 files at r1, 1 of 1 files at r2, 8 of 8 files at r3, 7 of 7 files at r4, 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MNYfWRz2NICZwpHFErZ:b-1oup4l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/kv/kvserver/observedts/uncertainty.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYaEL2EezcDAksZoJ-:-MNYaEL2EezcDAksZoJ0:bpt42j3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/kv/kvserver/observedts/uncertainty.go#L26)):*
> ```Go
> //
> // NOTE: if observedMaxTs is not empty, it must be <= maxTs.
> func IsUncertain(maxTs, observedMaxTs, valueTs hlc.Timestamp) bool {
> ```

Name bikeshedding thoughts: While we're considering a rename here, we may also want to ditch the "MaxTimestamp" nomenclature altogether. It doesn't tell the uninitiated anything! This is the upper boundary of the uncertainty interval. This is too much of a mouthful for a field name, but how about `UncertaintyTS`?

We could then name the adjusted value `OrganicUncertaintyTS`("organic" being the opposite of "synthetic"). The idea being, it's better to use this one, but you can only use it when encountering an organic, i.e. non-synthetic, timestamp.
If that doesn't resonate, `AdjustedUncertaintyTS` would be another option, or `ReducedUncertaintyTS`. Or extra descriptive with `ReducedOrganicUncertaintyTS`.
___
*[pkg/roachpb/data.go, line 1476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXABm-KXuJHbXfrDC:-MNYXABn9usB-xfNj-PM:bv9sjdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.go#L1476)):*
> ```Go
> 	// timestamp from the node.
> 	//
> 	// TODO(nvanbenschoten): how is this supposed to work for follower reads?
> ```

Follower reads don't hit that today, but would hit it with non-blocking txns, right?
___
*[pkg/roachpb/data.proto, line 342 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXOzCDYPAQvrrJe0g:-MNYXOzCDYPAQvrrJe0h:b95d69b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L342)):*
> ```protobuf
>   //
>   // Reads which encounter values with timestamps between read_timestamp and
>   // max_timestamp ("within the uncertainty interval") trigger a retry error
> ```

while you're massaging this, comment on whether the endpoints are inclusive.
___
*[pkg/roachpb/data.proto, line 357 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXbLy0xnC0c1hWzKA:-MNYXbLy0xnC0c1hWzKB:b-hkkkiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L357)):*
> ```protobuf
>   // been pushed; in this case, max_timestamp should be ignored.
>   util.hlc.Timestamp max_timestamp = 7 [(gogoproto.nullable) = false];
>   // The transaction's observed max timestamp is its max timestamp, reduce by
> ```

reduced
___
*[pkg/roachpb/data.proto, line 361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOy_Nl2MAUoiSeqWh4:-MNYYF3W7sjZOzJLf2Ov:bejpymb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L361)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Will this field ever be set in a stored `Transaction` proto, or used in an RPC?
Or is this just a convenient place to avoid plumbing through an extra parameter across function calls?
</blockquote></details>

It's just a convenient place for plumbing. I'd wager that Nathan dislikes this just as much as I do but doesn't want to get in the business of plumbing an auxiliary struct all the way down. I wonder though if now is a good time to do it. The change to avoid mutating `MaxTimestamp` in place is good, but introducing a bogus field smells. I'm on the fence. If we wrap the batch in a (non-proto) struct that can hold additional fields, we'll catch some allocations. Sure, we can amortize those easily over a `sync.Pool`, but still.
___
*[pkg/roachpb/data.proto, line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXhJaDovHkOCl4gKG:-MNYXhJaDovHkOCl4gKH:br5maty) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L366)):*
> ```protobuf
>   // most values on a range. This can lead to values that would otherwise be
>   // considered uncertain by the original max_timestamp to be considered
>   // "certainly concurrent" with the transaction due to observed timestamps.
> ```

, and thus not causally related.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 279 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYcNXl98krQin7g4iR:-MNYcNXl98krQin7g4iS:bxpcina) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L279)):*
> ```Go
> 
> // Returns whether a value with the specified timestamp is within the
> // transaction's uncertainty interval and is considered uncertainty.
> ```

considered uncertain
___
*[pkg/storage/pebble_mvcc_scanner.go, line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYd-py1hEO56OatLcG:-MNYd-py1hEO56OatLcH:b-k51bys) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L397)):*
> ```Go
> 		// Note that if we own the intent (i.e. we're reading transactionally)
> 		// we want to read the intent regardless of our read timestamp and fall
> 		// into case 8 below.
> ```

also the case 8 reference is confusing because this is case 8.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPQ2iW6yTT6XxrkSZw:-MNYeVxRCXjkwxV3UnQU:b66nscb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L397)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Since !otherIntentVisible, we have `!p.checkUncertainty || !p.isUncertainValue(metaTS)`. I am trying to understand this if-block. It would be executed when `!p.isUncertainValue(metaTS)`. Ah, I guess there could be committed values that are `< MaxTimestamp`. Could you expand a bit on the code comment to explain this.
</blockquote></details>

:heavy_plus_sign: 
___
*[pkg/storage/pebble_mvcc_scanner.go, line 656 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPSC8y606x0t_RAN7-:-MNYeTjP8He943ekEaaJ:b-7k8bwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L656)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

IIUC, this is needed because not everything in the interval (p.ts, seekTS] is an uncertainty error, since are doing seeks based on the worst-case uncertainty, because we need to look at the synthetic timestamp bit on what we find, to determine uncertainty. Could you add a longer comment here.

The loop below could also use a comment.
</blockquote></details>

:heavy_plus_sign: I'm pretty sure any time spent on these comments will pay off handsomely for future generations


<!-- Sent from Reviewable.io -->

nvb committed:
storage: introduce unused LocalUncertaintyLimit to MVCC

This commit introduces a new LocalUncertaintyLimit field on
MVCCScanOptions and MVCCGetOptions. The field is currently unused, but
will be tied into MVCC logic. To support this, we introduce a new test
that will be updated once the field begins to be used.

nvb committed:
storage: teach MVCC about local uncertainty limit and synthetic timestamps

This commit updates the `pebbleMVCCScanner` to understand the new
`LocalUncertaintyLimit` field and its interaction with synthetic
timestamps. These interactions make the determination of when a value is
uncertain more complex, as it creates situations where some values in a
transaction's full uncertainty interval are uncertain while others
aren't. This requires forward iteration in a few new places.

To centralize this logic, we pull it into a new `observedts.IsUncertain`
helper, which will also be used by the lock table.

To test this out, the commit updates the corresponding mvcc_histories
testdata file.

nvb committed:
roachpb: include LocalUncertaintyLimit in ReadWithinUncertaintyIntervalError

This commit adds the new LocalUncertaintyLimit value to the
ReadWithinUncertaintyIntervalError proto.

nvb committed:
kv: stop modifying MaxTimestamp, plumb local uncertainty limit

This commit replaces `LimitTxnMaxTimestamp` with `ComputeLocalUncertaintyLimit`.
Instead of updating the transaction proto's MaxTimestamp field, we now keep a
localUncertaintyLimit value on the stack during evaluation and use this to limit
uncertainty restarts. This makes things easier to understand, as we no longer
modify a request's transaction in-place.

In doing so, this also fixes a bug where an `mvccPutInternal` could return a
`ReadWithinUncertaintyInterval` error in cases where it should have returned a
`WriteTooOld` error.

nvb committed:
storage: demonstrate RWUI error case that should return WTO err

Will be fixed in next commit. This is important because we can often
refresh away a WriteTooOld error on the server, but we don't do as good
of a job refreshing away ReadWithinUncertaintyInterval errors on the
server. It's also a more consistent error to return in this situation.

nvb committed:
roachpb: rename Txn.MaxTimestamp to Txn.GlobalUncertaintyLimit

This commit renames Txn.MaxTimestamp to Txn.GlobalUncertaintyLimit, to
parallel the existing localUncertaintyLimit. The commit updates various
commentary to reflect this change.

nvb reviewed (commented):
TFTRs! I've pretty dramatically changed this PR in a few ways that I think people will like. But because of that, this is now based on #59086, so please hold off on looking at this again until that PR is further along.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MRbwAJME-8eZ-5bbIhS:busdsdq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[pkg/kv/kvserver/observedts/uncertainty.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYaEL2EezcDAksZoJ-:-MRbmy9e9jZ2fAmFlFpg:b-x0k6h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/kv/kvserver/observedts/uncertainty.go#L26)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Name bikeshedding thoughts: While we're considering a rename here, we may also want to ditch the "MaxTimestamp" nomenclature altogether. It doesn't tell the uninitiated anything! This is the upper boundary of the uncertainty interval. This is too much of a mouthful for a field name, but how about `UncertaintyTS`?

We could then name the adjusted value `OrganicUncertaintyTS`("organic" being the opposite of "synthetic"). The idea being, it's better to use this one, but you can only use it when encountering an organic, i.e. non-synthetic, timestamp.
If that doesn't resonate, `AdjustedUncertaintyTS` would be another option, or `ReducedUncertaintyTS`. Or extra descriptive with `ReducedOrganicUncertaintyTS`.
</blockquote></details>

Agreed. I switched to "local uncertainty limit" and "global uncertainty limit". Global indicates that the uncertainty limit applies globally while local, which is now only ever stored on the stack, indicates that the more aggressive uncertainty limit applies to the local node.
___
*[pkg/kv/kvserver/observedts/uncertainty.go, line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOzG9i80vR728RIz-5:-MRbmxBV9b2puZLvt5vv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b2b88d7ef9f6de83dff0a61efe62bab43814dc/pkg/kv/kvserver/observedts/uncertainty.go#L15)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

determines
</blockquote></details>

Done.
___
*[pkg/roachpb/data.go, line 1476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXABm-KXuJHbXfrDC:-MNe4lae6K2KILyuflKv:b-j3hjlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.go#L1476)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Follower reads don't hit that today, but would hit it with non-blocking txns, right?
</blockquote></details>

I think it is possible for follower reads to hit this (see [this logic](https://github.com/cockroachdb/cockroach/blob/e51e04cfa409ee2743c808b02d64d33df77dc07a/pkg/kv/kvserver/replica_follower_read.go#L70-L72)) since we allow long-running read-only transactions to begin serving reads from followers as soon as we determine that the closed timestamp is sufficiently high. I'm guessing we've just never seen this in the wild because the transaction needs to be fairly long-running.

But you're right, we'll need to address this for non-blocking txns. This is tracked in #57685.
___
*[pkg/roachpb/data.proto, line 342 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXOzCDYPAQvrrJe0g:-MRboJjzAwUimJITn1xa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L342)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

while you're massaging this, comment on whether the endpoints are inclusive.
</blockquote></details>

Done.
___
*[pkg/roachpb/data.proto, line 357 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXbLy0xnC0c1hWzKA:-MNe6H8m-YsuUIAy64TW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L357)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

reduced
</blockquote></details>

Done.
___
*[pkg/roachpb/data.proto, line 361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOy_Nl2MAUoiSeqWh4:-MNe6Q0U3fIGrtLycOJz:bals0tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L361)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It's just a convenient place for plumbing. I'd wager that Nathan dislikes this just as much as I do but doesn't want to get in the business of plumbing an auxiliary struct all the way down. I wonder though if now is a good time to do it. The change to avoid mutating `MaxTimestamp` in place is good, but introducing a bogus field smells. I'm on the fence. If we wrap the batch in a (non-proto) struct that can hold additional fields, we'll catch some allocations. Sure, we can amortize those easily over a `sync.Pool`, but still.
</blockquote></details>

Done with the help of #59086. How do you two like this approach?
___
*[pkg/roachpb/data.proto, line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYXhJaDovHkOCl4gKG:-MRbp0PT-uRaRakiW-_U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L366)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

, and thus not causally related.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 656 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOxv0e8obXAfymg7V1:-MRXz7vEA-HDq62OtoR8:b3spg7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ef06227291b6e1ac80cc595db06cefff2b3c5c5/pkg/storage/pebble_mvcc_scanner.go#L656)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is it worth keeping this check as an assertion, in case there is data corruption?
</blockquote></details>

I wasn't thinking so. There are plenty of other invariants in our MVCC encoding that we assume hold throughout this code, so I don't think it's worth it to keep a small bit of defensive code if it obscures clarity.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 279 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYcNXl98krQin7g4iR:-MRbq1ph946vCu97KUa8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L279)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

considered uncertain
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPQ2iW6yTT6XxrkSZw:-MRbqC9zCbC26ETrrjul:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L397)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

:heavy_plus_sign: 
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNYd-py1hEO56OatLcG:-MRbqe4p5Tjgwi1vs3oA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L397)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

also the case 8 reference is confusing because this is case 8.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 656 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNPSC8y606x0t_RAN7-:-MRbsTT_AjbeR8gFRaMO:b-vkx5g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07110aa5e862f3726d6f6d7f1ca2c3f1dc34a4/pkg/storage/pebble_mvcc_scanner.go#L656)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

:heavy_plus_sign: I'm pretty sure any time spent on these comments will pay off handsomely for future generations
</blockquote></details>

Done. I got a little carried away trying to maintain the terseness of this code.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
I looked through some parts of this, including the change to `pebble_mvcc_scanner.go` (I'm unable to get reviewable to behave reasonably with many commits). I'll do a more careful pass tomorrow.

> Reviewed 1 of 22 files at r6, 1 of 1 files at r9, 1 of 8 files at r10, 1 of 8 files at r11, 2 of 61 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MSVI5Q_7hFFQC4mCC7n:bch73pg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[pkg/storage/mvcc.go, line 1203 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSVDwCoD20EXHpeNqMf:-MSVDwCoD20EXHpeNqMg:b-amzavw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/storage/mvcc.go#L1203)):*
> ```Go
> // behalf of a transaction), will result in a WriteIntentError. Because
> // of this, the function is only called from places where intents have
> // already been considered.
> ```

Was this "places where intents have already been considered" always the case, or is this new here?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> I'll do a more careful pass tomorrow.

Finished a full pass -- only some small comments in addition to my earlier question in mvcc.go.

> Reviewed 3 of 8 files at r11, 10 of 18 files at r12, 9 of 18 files at r14, 28 of 61 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MSdW9gfEh36tgU8-UeP:bmfixj4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/observedts/doc.go, line 152 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSoa150kq6Scp4f6I:-MSdSoa150kq6Scp4f6J:b-9r2mkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L152)):*
> ```Go
> // of the lease that the request is executing under before using it to limit the
> // request's uncertainty interval.
> //
> ```

Should there be an addition to this comment to account for https://github.com/cockroachdb/cockroach/issues/36431
___
*[pkg/kv/kvserver/observedts/doc.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSgkV9bm36j_DjyKN:-MSdSgkV9bm36j_DjyKO:bowxinj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L173)):*
> ```Go
> // being served from a different replica in the range, the observed timestamp
> // still places a bound on the values in the range that may have been written
> // before the transaction began.
> ```

It would also need to know when the leaseholder acquired the lease for the range, yes?
___
*[pkg/roachpb/data.go, line 1486 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdUFqU241RqV_jfqLh:-MSdUFqU241RqV_jfqLi:bbwfkj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/roachpb/data.go#L1486)):*
> ```Go
> 	// This is tracked in #57685. We can now use the LocalUncertaintyLimit in
> 	// place of clockTS, which handles follower reads correctly.
> 	clockTS, ok := txn.GetObservedTimestamp(origin)
> ```

If the uncertainty was caused by the LocalUncertaintyLimit being adjusted by the lease acquisition time (which I think is not reflected in GetObservedTimestamp, since it is specific to a replica and not the node) won't this time not be high enough to remove the uncertainty? If so, will it keep retrying at this low timestamp indefinitely?
___
*[pkg/roachpb/data.proto, line 361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MNOy_Nl2MAUoiSeqWh4:-MSdUwJ1CUesAKzc8PrV:b-47fafm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3197a2cda293da7fc3919cbc9c655a98203253e/pkg/roachpb/data.proto#L361)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done with the help of #59086. How do you two like this approach?
</blockquote></details>

looks good


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MT3_wuS6f3kLFMTPKtZ:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[pkg/kv/kvserver/observedts/doc.go, line 152 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSoa150kq6Scp4f6I:-MT3XK64E_MdkwJ55rie:b-wrvmuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L152)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Should there be an addition to this comment to account for https://github.com/cockroachdb/cockroach/issues/36431
</blockquote></details>

Yes, absolutely. I was planning on adding that once we fix #36431 by setting the synthetic flag when intents are moved during intent resolution.
___
*[pkg/kv/kvserver/observedts/doc.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSgkV9bm36j_DjyKN:-MT3XovfAKzEJL53u4La:b-ot07ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L173)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

It would also need to know when the leaseholder acquired the lease for the range, yes?
</blockquote></details>

Could you expand on this? The follower knows the lease start time, if that's what you mean.
___
*[pkg/roachpb/data.go, line 1486 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdUFqU241RqV_jfqLh:-MT3YQDcD2_xhm-2eyMf:bdkm20y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/roachpb/data.go#L1486)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

If the uncertainty was caused by the LocalUncertaintyLimit being adjusted by the lease acquisition time (which I think is not reflected in GetObservedTimestamp, since it is specific to a replica and not the node) won't this time not be high enough to remove the uncertainty? If so, will it keep retrying at this low timestamp indefinitely?
</blockquote></details>

I think this case is handled by the forward to `err.ExistingTimestamp`. In theory, that's all we need to bump our timestamp to in order to ensure that we observe the uncertain value after a restart and avoid an infinite retry loop. I guess we decided in https://github.com/cockroachdb/cockroach/commit/9eb55b71504a5da0670d4fc5b7a727a7bf975c63#diff-230152fb9c39bac311935d3cda2ec1d43791a94de9316a1366b314b21337f51eR735 that we might as well also forward the restart time up to the observed timestamp to avoid any other uncertainty restarts on this node.

I'm tracking this in https://github.com/cockroachdb/cockroach/issues/57685, so I'll clean this up and provide a better comment when I address that sometime in the next two weeks.
___
*[pkg/storage/mvcc.go, line 1203 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSVDwCoD20EXHpeNqMf:-MT3X33EC7onYSin0G4B:b1xqbsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/storage/mvcc.go#L1203)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Was this "places where intents have already been considered" always the case, or is this new here?
</blockquote></details>

This was already the case, but we didn't call that out anywhere.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
Looks good to me, but I'm not well versed in some of the KV code.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 31 files at r1, 5 of 22 files at r6, 9 of 20 files at r7, 9 of 33 files at r8, 11 of 97 files at r16, 3 of 8 files at r18, 7 of 18 files at r19, 12 of 18 files at r21, 59 of 61 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MTBbjc7ByHgHOi6415O:bddoq2k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/batcheval/cmd_lease_transfer.go, line 36 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MTB_ACw7z8XqeZ8rVYw:-MTB_ACw7z8XqeZ8rVYx:b-dobh4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/batcheval/cmd_lease_transfer.go#L36)):*
> ```Go
> 	//
> 	// Because of this, it declares a non-MVCC write over every addressable key
> 	// in the range, even through the only key the TransferLease actually writes
> ```

s/through/though/
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 127 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MTB_oMf5Y0d0I3I8h4D:-MTB_oMf5Y0d0I3I8h4E:b-6p09jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L127)):*
> ```Go
> 	}
> 	now := cArgs.EvalCtx.Clock().Now()
> 	// TODO(nvanbenschoten): remove this limitation. But when doing so,
> ```

 what is "this limitation"?
___
*[pkg/kv/kvserver/observedts/doc.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSgkV9bm36j_DjyKN:-MTBQcBZ4Pgbs0OmGMHc:b7s4bet) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L173)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you expand on this? The follower knows the lease start time, if that's what you mean.
</blockquote></details>

Yes, that's what I meant -- otherwise we could set the local uncertainty limit too low for follower reads. Does the follower maintain the history of lease start times, or it will just use the current leaseholder. I am assuming the latter is sufficient, and it seems that is what `limit.go` is doing. Some crude reasoning to convince myself (it would be nice to have a refined sketch of such reasoning as a code comment in `limit.go`): 
T_txn: transaction time
T_lease_start: leaseholder's lease start time
T_closed_time: closed time from that leaseholder
T_txn <= T_closed_time to do follower reads
If T_txn >= T_lease_start, I think the correctness is straightforward.
However, it is possible that T_lease_start > T_txn, and that the txn had an observed timestamp from the earlier leaseholder. But we are safe because T_lease_start > than that observed timestamp from the earlier leaseholder.

Is this correct?
___
*[pkg/roachpb/data.go, line 1486 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdUFqU241RqV_jfqLh:-MTBOANSBjanIM3mSSjq:b-wpehnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/roachpb/data.go#L1486)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think this case is handled by the forward to `err.ExistingTimestamp`. In theory, that's all we need to bump our timestamp to in order to ensure that we observe the uncertain value after a restart and avoid an infinite retry loop. I guess we decided in https://github.com/cockroachdb/cockroach/commit/9eb55b71504a5da0670d4fc5b7a727a7bf975c63#diff-230152fb9c39bac311935d3cda2ec1d43791a94de9316a1366b314b21337f51eR735 that we might as well also forward the restart time up to the observed timestamp to avoid any other uncertainty restarts on this node.

I'm tracking this in https://github.com/cockroachdb/cockroach/issues/57685, so I'll clean this up and provide a better comment when I address that sometime in the next two weeks.
</blockquote></details>

(just replaying what you said to make sure I understood it right)
We expect the output of `txn.GetObservedTimestamp(origin)` to be typically `> err.ExistingTimestamp`, except when the local uncertainty was bumped up due to lease acquisition, where the latter could be higher. Either way we are increasing the txn timestamp beyond the write seen within the uncertainty limit, which ensures liveness.

A code comment here would be helpful.

My comment here was not about follower reads, but I'm assuming you mean that follower reads need improving here because the origin is the follower node and we don't want the follower node's observed timestamp. Correct?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-:-MTEJ4UB7TKFCDlrbaCO:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @tbg)

___
*[pkg/kv/kvserver/batcheval/cmd_lease_transfer.go, line 36 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MTB_ACw7z8XqeZ8rVYw:-MTEFiooCQJyS-ueCpBh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/batcheval/cmd_lease_transfer.go#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/through/though/
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 127 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MTB_oMf5Y0d0I3I8h4D:-MTEFkn241bNSVSr0uhi:b-6ctxwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L127)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

 what is "this limitation"?
</blockquote></details>

The limitation was that transactions could not be pushed into the future with synthetic timestamps. This will be fixed by https://github.com/cockroachdb/cockroach/pull/59693.
___
*[pkg/kv/kvserver/observedts/doc.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdSgkV9bm36j_DjyKN:-MTEG2x1A7vyExhAFrZn:b2gxlsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/kv/kvserver/observedts/doc.go#L173)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Yes, that's what I meant -- otherwise we could set the local uncertainty limit too low for follower reads. Does the follower maintain the history of lease start times, or it will just use the current leaseholder. I am assuming the latter is sufficient, and it seems that is what `limit.go` is doing. Some crude reasoning to convince myself (it would be nice to have a refined sketch of such reasoning as a code comment in `limit.go`): 
T_txn: transaction time
T_lease_start: leaseholder's lease start time
T_closed_time: closed time from that leaseholder
T_txn <= T_closed_time to do follower reads
If T_txn >= T_lease_start, I think the correctness is straightforward.
However, it is possible that T_lease_start > T_txn, and that the txn had an observed timestamp from the earlier leaseholder. But we are safe because T_lease_start > than that observed timestamp from the earlier leaseholder.

Is this correct?

</blockquote></details>

Just the current leaseholder. This code is actually agnostic to whether the read is off a leaseholder or a follower. I agree that we could discuss this more in the comment in `observedts/doc.go`'s `Follower Reads` section. I'll do that when I go back to update that comment.
___
*[pkg/roachpb/data.go, line 1486 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/57077#-MSdUFqU241RqV_jfqLh:-MTEI3JR3XYdCgvRoqgr:b1w90a0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8c5ffeee3823a06dc1d9d3ad35dd8f53f7249ea/pkg/roachpb/data.go#L1486)):*
> We expect the output of txn.GetObservedTimestamp(origin) to be typically > err.ExistingTimestamp, except when the local uncertainty was bumped up due to lease acquisition, where the latter could be higher. Either way we are increasing the txn timestamp beyond the write seen within the uncertainty limit, which ensures liveness.

Yes, this all sounds correct.

> A code comment here would be helpful.

Agreed. We can discuss what more to add in https://github.com/cockroachdb/cockroach/pull/60021. The comment in there says a lot of this, but it may be missing aspects.

> My comment here was not about follower reads, but I'm assuming you mean that follower reads need improving here because the origin is the follower node and we don't want the follower node's observed timestamp. Correct?

That was my original thinking, but I now think I was wrong. We will have an observed timestamp from the follower node, but it won't be relevant to the value because it can't be applied to establish a local uncertainty limit. So https://github.com/cockroachdb/cockroach/pull/60021 might not actually be necessary, but it seems like a good improvement regardless.


<!-- Sent from Reviewable.io -->

nvb committed:
kv: fix flaky timestamp assertion in TestRangeLocalUncertaintyLimitAfterNewLease

There is no guarantee that the receiving end of a lease transfer will
have a high clock than the sending end after the lease transfer
completes. There is only a guarantee that the receiving end will have a
higher clock than the sending end had before the transfer began.

This is likely fallout from the recent switch to the use of
HybridManualClock in this test, but showed up in CI on this PR because
the test rename triggered CI to run it under extra stress.

nvanbenschoten commented:
bors r+
